src/storage/src/source/sql_server/progress.rs

```diff
@@ -0,0 +1,186 @@
+                .resolve_config(
+                    &config.config.connection_context.secrets_reader,
+                    &config.config,
+                    InTask::Yes,
+                )
+                .await?;
+            let (mut client, conn) = mz_sql_server_util::Client::connect(conn_config).await?;
+            // TODO(sql_server1): Move the connection into its own future.
+            mz_ore::task::spawn(|| "sql_server-progress-conn", async move { conn.await });
+
+            let probe_interval = OFFSET_KNOWN_INTERVAL.handle(config.config.config_set());
+            let mut probe_ticker = probe::Ticker::new(|| probe_interval.get(), config.now_fn);
+
+            // Offset that is measured from the upstream SQL Server instance.
+            let mut prev_offset_known: Option<Lsn> = None;
+            // Offset that we have observed from the `resume_uppers` stream.
+            let mut prev_offset_committed: Option<Lsn> = None;
+
+            // This stream of "resume uppers" tracks all of the Lsn's that we have durably
+            // committed for all subsources/exports and thus we can notify the upstream that the
+            // change tables can be cleaned up.
+            let mut resume_uppers = std::pin::pin!(resume_uppers);
+            let cleanup_change_table = CDC_CLEANUP_CHANGE_TABLE.handle(config.config.config_set());
+            let cleanup_max_deletes = CDC_CLEANUP_CHANGE_TABLE_MAX_DELETES.handle(config.config.config_set());
+            let capture_instances: BTreeSet<_> = outputs.into_values().map(|info| info.capture_instance).collect();
+
+            loop {
+                tokio::select! {
+                    // TODO(sql_server2): Emit probes here.
+                    _probe_ts = probe_ticker.tick() => {
+                        let known_lsn = mz_sql_server_util::inspect::get_max_lsn(&mut client).await?;
+
+                        // The DB should never go backwards, but it's good to know if it does.
+                        let prev_known_lsn = match prev_offset_known {
+                            None => {
+                                prev_offset_known = Some(known_lsn);
+                                known_lsn
+                            },
+                            Some(prev) => prev,
+                        };
+                        if known_lsn < prev_known_lsn {
+                            mz_ore::soft_panic_or_log!(
+                                "upstream SQL Server went backwards in time, current LSN: {known_lsn}, last known {prev_known_lsn}",
+                            );
+                            continue;
+                        }
+
+                        // Update any listeners with our most recently known LSN.
+                        if let Some(prev_commit_lsn) = prev_offset_committed {
+                            let known_lsn_abrv = known_lsn.abbreviate();
+                            let commit_lsn_abrv = prev_commit_lsn.abbreviate();
```

Comment:
What kind of precision loss are we suffering here by abbreviating?

Reply:
We lose the `record_id` field which denotes an individual operation within a log block, but AFAICT (based on the [docs](https://learn.microsoft.com/en-us/sql/relational-databases/sql-server-transaction-log-architecture-and-management-guide?view=sql-server-ver16#log-blocks) and experimentation) the `(vlf_id, log_block)` is the identifier of a single transaction, while the `record_id` denotes the the individual updates within said transaction

Reply:
Ah, cool, correlating with the number of transactions sound great